doc/7/getting-started/.react/standalone/src/App.js

```diff
@@ -0,0 +1,165 @@
+/* snippet:start:1 */
+import React from 'react';
+import kuzzle from "./services/kuzzle";
+import './App.css';
+/* snippet:end */
+/* snippet:start:8 */
+const Message = function (props) {
+  const { message, username } = props;
+  return (
+    <div className={(message.username === username ? 'fromMe' : 'fromOthers') + ' messages'}>
+      <span> User: <b>{message.username}</b>  (</span>
+      <span> {new Date(message.createdAt).toLocaleString().split("GMT")[0]} )</span>
+      <p> {message.value} </p>
+    </div>
+  );
+}
+/* snippet:end */
+/* snippet:start:2 */
+class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      username: "",
+      message: "",
+      messages: [],
+      validate: false,
+    };
+  }
+/* snippet:end */
+  /* snippet:start:3 */
+  valid = async () => {
+    await kuzzle.connect();
+    if (!await kuzzle.index.exists("chat")) {
+      await kuzzle.index.create("chat");
+      await kuzzle.collection.create("chat", "messages");
+    }
+    await this.subscribeMessages();
+    await this.fetchMessages();
+    this.setState({ validate: true });
+  }
+  /* snippet:end */
+  /* snippet:start:10 */
+  subscribeMessages = () => {
+    return (
+      kuzzle.realtime.subscribe(
+        "chat",
+        "messages", {},
+        notification => {
+          if (notification.type !== "document") return;
+          if (notification.action !== "create") return;
+          this.setState({
+            messages: [
+              this.getMessage(notification.result),
+              ...this.state.messages
+            ]
+          })
+        }
+      )
+    );
+  }
+  /* snippet:end */
+  /* snippet:start:6 */
+  getMessage = document => {
+    const message = {
+      _id: document._id,
+      value: document._source.value,
+      createdAt: document._source._kuzzle_info.createdAt,
+      username: document._source.username
+    };
+    return message;
+  }
+  /* snippet:end */
+  /* snippet:start:7 */
+  fetchMessages = async () => {
+    const results = await kuzzle.document.search(
+      "chat",
+      "messages",
+      {
+        sort: ["_kuzzle_info.createdAt"]
+      },
+      {
+        size: 100
+      }
+    );
+    let messages = [];
+    results.hits.forEach(hit => {
+      messages = [this.getMessage(hit), ...messages]
```

Comment:
For each hit, the JS runtime will:
 - creating a new array with the new message and the old ones
 - assigning this new array to `messages`
 - garbage collect the old `messages` array

This is quite inefficient and you should simply use `Array.push` to push back an element to the end of the array.

```
concat & assign    x 536,959 ops/sec ±1.35% (88 runs sampled)
push   x 7,636,825 ops/sec ±0.63% (92 runs sampled)
```